crypto: Switch to base58 for TransactionDigest

joyqvq opened the pull request.

666lcz commented:
Can we also update https://github.com/MystenLabs/sui/blob/21c26ce6a5d4e3448abd74323e3164286d3deba6/sdk/typescript/src/types/common.ts#L23 ?

joyqvq commented:
> Can we also update
> 
> https://github.com/MystenLabs/sui/blob/21c26ce6a5d4e3448abd74323e3164286d3deba6/sdk/typescript/src/types/common.ts#L23
> 
> ?

added - there is no e2e search test for tx digest thats why nothing fails on ci. just added a e2e test for that

kchalkias reviewed (commented):
Changes seem fine, are we sure nothing else breaks? wallet, capy, and others that need to be informed? @damirka ?

damirka commented:
Thanks for pinging me @kchalkias! Is the change only for tx digest (and not object digest)? If so, then we’re good. If it affects object digest then we’ll need changes in the bcs implementation as well.

joyqvq commented:
> Thanks for pinging me @kchalkias! Is the change only for tx digest (and not object digest)? If so, then we’re good. If it affects object digest then we’ll need changes in the bcs implementation as well.

correct not object digest. although it should not impact bcs regardless, since bcs is concerning struct -> bytes not bytes -> string, base58 would only affect the latter.

Jordan-Mysten commented on sdk/typescript/src/cryptography/hash.ts:
```diff
@@ -20,5 +20,5 @@ export function sha256Hash(typeTag: string, data: Uint8Array): string {
   hash.update(dataWithTag);
 
-  return Buffer.from(hash.hex(), 'hex').toString('base64');
+  return Buffer.from(hash.hex(), 'hex');
```
When did we re-add buffer to the SDK? We should never use the native Buffer type.

joyqvq replied:
good catch, this was recently introduced, removed.

Jordan-Mysten commented on sdk/typescript/src/types/common.ts:
```diff
@@ -13,8 +13,9 @@ import {
+import { Base64DataBuffer } from '../serialization/base64';
 
-/** Base64 string representing the object digest */
+/** Base58 string representing the transaction digest */
```
We should support both at the same time while this rolls out.

joyqvq replied:
good call. in the case of explorer gets deployed before sui, it should parse base64.

joyqvq commented on apps/explorer/src/utils/static/mock_data.json:
```diff
@@ -595,7 +595,7 @@
             "readonly": "true",
             "objType": "0x2::coin::Coin<0x2::USD::USD>",
             "data": {
-                "tx_digest": "Da4vHc9IwbvOYblE8LnrVsqXwryt2Kmms+xnJ7Zx5E4=",
```
mock_data will be deprecated in favor of localnet e2e test, still added here for consistency

joyqvq commented on sdk/typescript/src/cryptography/hash.ts:
```diff
@@ -20,5 +21,5 @@ export function sha256Hash(typeTag: string, data: Uint8Array): string {
 
   hash.update(dataWithTag);
 
-  return Buffer.from(hash.hex(), 'hex').toString('base64');
```
buffer removed.

666lcz commented on sdk/typescript/src/types/common.ts:
```diff
@@ -26,17 +26,23 @@ export type ObjectOwner =
+    buffer = new Base58DataBuffer(value);
+  } catch (e) {
+    try {
+      // For backwards compatibility, we also accept Base64 strings
```
I would pass in a `serializationFmt: 'base64' | 'base58'` to this method. The application will choose which one to use depending on API version

joyqvq replied:
revised

666lcz commented on sdk/typescript/src/types/common.ts:
```diff
@@ -134,7 +140,8 @@ export function generateTransactionDigest(
     .toBytes();
-  return sha256Hash('SenderSignedData', senderSignedDataBytes);
+
+  return new Base58DataBuffer(sha256Hash('SenderSignedData', senderSignedDataBytes)).toString();
```
For backward compatibility, I think you can pass in a `serializationFmt: 'base58' | 'base64' ` to the function signature of `generateTransactionDigest`. And in line 144, you can do `serializationFmt === 'base58' ? ... : ...)

joyqvq replied:
revised

666lcz commented:
Looks like there's some build error to fix - otherwise looks great to me! Thanks for adding the tests!

kchalkias reviewed (approved):
Agree for backwards b64 compatibility. I guess our explorer and wallets (and partners gradually) will prefer the new b58 one and the teams need to know.
